### PR TITLE
CORE-4868 - update to latest corda api version with cpk_metadata PKs

### PR DESCRIPTION
**Pull request by conal-smith**
API change: https://github.com/corda/corda-api/pull/421

### CODE REVIEW

File: components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt [hunk 1]
```diff
@@ -205,11 +207,16 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
...
-            it.find(CpkFileEntity::class.java, cpkChecksum.toString())
+        val query = "SELECT count(c) FROM ${CpkFileEntity::class.simpleName} c WHERE c.fileChecksum = :cpkFileChecksum"
+        val entitiesFound = entityManagerFactory.createEntityManager().transaction {
+            it.createQuery(query)
+                .setParameter("cpkFileChecksum", cpkChecksum.toString())
+                .singleResult as Long
         }
 
-        return entity != null
+        if (entitiesFound > 1) throw NonUniqueResultException("CpkFileEntity with fileChecksum = $cpkChecksum was not unique")
```
Comment (sprintengine) [inline components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:217]: This check may not be necessary as the column already has unique constraint at the database level.

File: components/virtual-node/cpk-write-service-impl/src/main/kotlin/net/corda/cpk/write/impl/services/db/impl/DBCpkStorage.kt [hunk 2]
```diff
@@ -38,10 +38,13 @@ class DBCpkStorage(private val entityManagerFactory: EntityManagerFactory) : Cpk
...
-            val cpkDataEntity = it.find(
-                CpkFileEntity::class.java,
-                checksum.toString()
+            val cpkDataEntity = it.createQuery(
+                "FROM ${CpkFileEntity::class.java.simpleName} WHERE fileChecksum = :checksum",
+                CpkFileEntity::class.java
             )
+                .setParameter("checksum", checksum.toString())
+                .singleResult
+
```
Comment (sprintengine) [inline components/virtual-node/cpk-write-service-impl/src/main/kotlin/net/corda/cpk/write/impl/services/db/impl/DBCpkStorage.kt:47]: We can still utilize the file checksum which is indexed on CpkFileEntity with unique constraint

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt [hunk 3]
```diff
@@ -60,6 +64,10 @@ data class CpiCpkKey(
     private val cpiVersion: String,
     @Column(name = "cpi_signer_summary_hash")
     private val cpiSignerSummaryHash: String,
-    @Column(name = "cpk_file_checksum")
-    private val cpkFileChecksum: String,
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt:64]: CpiCpk now uses cpkName, version and signer summary hash as FK to cpk_metadata not the cpkFileChecksum

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt [hunk 4]
```diff
@@ -402,38 +401,65 @@ internal class DatabaseChunkPersistenceTest {
...
+        val inASet = setOf(
+            s1.toString(),
+            s2.toString(),
+            s3.toString(),
+            s4.toString(),
+            s5.toString(),
+        )
+
+        assertThat(inASet).hasSize(5)
+    }
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt]: Is this a valuable test?

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt [hunk 5]
```diff
@@ -537,5 +616,14 @@ internal class DatabaseChunkPersistenceTest {
...
+        // merging updated cpi accounts for 1 modification + modifying cpk
+        assertThat(updatedCpi.entityVersion).isEqualTo(3)
+        assertThat(updatedCpi.cpks.size).isEqualTo(1)
+        assertThat(updatedCpi.cpks.first().entityVersion).isEqualTo(1)
+    }
+
+    @Test
+    fun `cannot persist two CPK metadatas with the same name, version and signer hash, even if checksum is different`() {
+
     }
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt]: Is the implementation of this test missing?

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt (hunk 4 above)
Comment (sprintengine) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt]: I think we're OK to remove this

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt (hunk 5 above)
Comment (sprintengine) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt]: Yes, good spot, I'll add this test now

Comment (sprintengine) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt]: I think `database chunk persistence can force update the same CPI` already makes the assertions this test was originally titled to do, i.e. persisting a cpi with a new cpk with the same ID will overwrite the original and increment the entityVersion of the cpk. So I think I'm happy enough to just remove this.

File: gradle.properties [hunk 6]
```diff
@@ -39,7 +39,7 @@ commonsTextVersion = 1.9
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
 #cordaApiVersion=5.0.0.xxx-SNAPSHOT
-cordaApiVersion=5.0.0.117-beta+
+cordaApiVersion=5.0.0.118-alpha-1655204096912
```
Comment (vkolomeyko) [inline gradle.properties]: I have merged `corda-api` PR, once beta version is built it should be possible to update to it.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt [hunk 7]
```diff
@@ -26,14 +26,18 @@ data class CpiCpkEntity(
...
+        JoinColumn(name = "cpk_version", referencedColumnName = "cpk_version", insertable = false, updatable = false),
         JoinColumn(
-            name = "cpk_file_checksum",
-            referencedColumnName = "file_checksum",
+            name = "cpk_signer_summary_hash",
+            referencedColumnName = "cpk_signer_summary_hash",
             insertable = false,
             updatable = false
-        )
+        ),
```
Comment (vkolomeyko) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt:40]: Minor: Trailing comma can be removed

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt [hunk 8]
```diff
@@ -60,6 +64,10 @@ data class CpiCpkKey(
...
     @Column(name = "cpi_signer_summary_hash")
     private val cpiSignerSummaryHash: String,
-    @Column(name = "cpk_file_checksum")
-    private val cpkFileChecksum: String,
+    @Column(name = "cpk_name")
+    private val cpkName: String,
+    @Column(name = "cpk_version")
+    private val cpkVersion: String,
+    @Column(name = "cpk_signer_summary_hash")
+    private val cpkSignerSummaryHash: String,
```
Comment (vkolomeyko) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt:72]: Minor: Trailing comma can be removed

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/virtualnode/VirtualNodeRpcTest.kt [hunk 9]
```diff
@@ -242,13 +241,14 @@ class VirtualNodeRpcTest {
 
             // Depends on the flows in the cpi
             val className = "net.cordapp.testing.calculator.CalculatorFlow"
+            val clientRequestId = Random.nextInt()
```
Comment (vkolomeyko) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/virtualnode/VirtualNodeRpcTest.kt:244]: Using `System.currentTimeMillis` might be a better option to make sure we do not repeat request Id between E2E tests runs.

**Review by vkolomeyko — DISMISSED** [review]
Overall LGTM - few minor remarks posted.

**Review by vkolomeyko — DISMISSED** [review]
LGTM

**Review by driessamyn — DISMISSED** [review]
LGTM

**Review by barrylapthorn — DISMISSED** [review]

**Review by kyriathar — APPROVED** [review]
LGTM

**Review by vkolomeyko — APPROVED** [review]

**Pull request merged**